File: algorithm/leetcode/322_coinChange.py
class Solution(object):
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        dp = [float('inf')]*(amount + 1)
        dp[0] = 0
        for coin in coins:
            for num in range(coin, amount + 1):
                dp[num] = min(dp[num], dp[num-coin] + 1)
        return dp[amount] if dp[amount] != float('inf') else -1

so = Solution()
res = so.coinChange([1,2,5], 11)

File: algorithm/leetcode/322_coinChange.cpp
#include <vector>

using namespace std;

class Solution
{
public:
    int coinChange(vector<int>& coins, int amount){
        vector<int> dp(amount + 1, INT16_MAX);
        dp[0] = 0;
        for(int i = 0; i < coins.size(); i++){
            for(int j = coins[i]; j < amount + 1; j++){
                dp[j] = min(dp[j], dp[j - coins[i]] + 1);
            }
        }
        return dp[amount] > amount ? -1 : dp[amount];
    }
};

int main(){
    Solution so = Solution();
    vector<int> coins = {1, 2, 5};
    int res = so.coinChange(coins, 11);
}


